modules/fedimint-mint-client/src/issuance.rs

```diff
@@ -0,0 +1,372 @@
+use std::time::Duration;
+
+use fedimint_client::sm::{ClientSMDatabaseTransaction, OperationId, State, StateTransition};
+use fedimint_client::DynGlobalClientContext;
+use fedimint_core::api::GlobalFederationApi;
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::module::registry::ModuleDecoderRegistry;
+use fedimint_core::{Amount, OutPoint, Tiered, TieredMulti, TransactionId};
+use fedimint_derive_secret::{ChildId, DerivableSecret};
+use fedimint_mint_common::{BlindNonce, MintOutputBlindSignatures, MintOutputOutcome, Nonce, Note};
+use secp256k1::{KeyPair, Secp256k1, Signing};
+use serde::{Deserialize, Serialize};
+use tbs::{blind_message, unblind_signature, AggregatePublicKey, BlindedSignature, BlindingKey};
+use thiserror::Error;
+use tracing::error;
+
+use crate::db::NoteKey;
+use crate::{MintClientContext, SpendableNote};
+
+/// State machine managing the e-cash issuance process related to a mint output.
+///
+/// ```mermaid
+/// graph LR
+///     classDef virtual fill:#fff,stroke-dasharray: 5 5
+///
+///     Created -- containing tx rejected --> Aborted
+///     Created -- await output outcome --> Outcome["Outcome Received"]:::virtual
+///     subgraph Await Outcome
+///     Outcome -- valid blind signatures  --> Succeeded
+///     Outcome -- invalid blind signatures  --> Failed
+///     end
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub enum MintIssuanceStates {
+    /// Issuance request was created, we are waiting for blind signatures
+    Created(MintIssuanceStatesCreated),
+    /// The transaction containing the issuance was rejected, we can stop
+    /// looking for preimages
+    Aborted(MintIssuanceStatesAborted),
```

Comment:
The rejected vs aborted seems weird. Should we just stick to one?

Reply:
Well, `Aborted` is kinda ok, we just get a refund, while `Failed` means we need to manually intervene to recover funds (if possible).